### PR TITLE
x11: fix --on-all-workspaces option

### PR DESCRIPTION
**Pull request by odnar-dev**
mpv set _NET_WM_DESKTOP to 0xFFFFFFFF, which behaves differently with each window manager. Instead, set the window property to STICKY, and let the window managers deal with it.

tested on `archlinux` with `xfce 4.16` and `bspwm 0.9.10`

### CODE REVIEW

**Comment by odnar-dev** [comment]
@Dudemanguy

**Review by Dudemanguy — APPROVED** [review]
LGTM

File: video/out/x11_common.c [hunk 1]
```diff
@@ -1606,10 +1609,16 @@ static void vo_x11_map_window(struct vo *vo, struct mp_rect rc)
     }
 
     if (x11->opts->all_workspaces || x11->opts->geometry.ws > 0) {
-        long v = x11->opts->all_workspaces
-                ? 0xFFFFFFFF : x11->opts->geometry.ws - 1;
-        XChangeProperty(x11->display, x11->window, XA(x11, _NET_WM_DESKTOP),
-                        XA_CARDINAL, 32, PropModeReplace, (unsigned char *)&v, 1);
+        if (x11->wm_type & vo_wm_STICKY) {
```
Comment (sfan5) [inline video/out/x11_common.c:1612]: this looks incorrect, `x11->opts->geometry.ws > 0` should still use _NET_WM_DESKTOP with the specific workspace

File: video/out/x11_common.c [hunk 2]
```diff
@@ -1606,10 +1609,16 @@ static void vo_x11_map_window(struct vo *vo, struct mp_rect rc)
...
+            Atom state = XA(x11, _NET_WM_STATE_STICKY);
+            XChangeProperty(x11->display, x11->window, XA(x11, _NET_WM_STATE), XA_ATOM,
+                            32, PropModeAppend, (unsigned char *)&state, 1);
```
Comment (sfan5) [inline video/out/x11_common.c:1615]: not that it matters but `PropModeReplace` maybe?

File: video/out/x11_common.c (hunk 1 above)
Comment (odnar-dev) [reply video/out/x11_common.c:1612]: i couldn't  find what `x11->opts->geometry.ws` is, so i left it as it is.
what about
```patch
---
 video/out/x11_common.c | 21 ++++++++++-----------
 1 file changed, 10 insertions(+), 11 deletions(-)

diff --git a/video/out/x11_common.c b/video/out/x11_common.c
index 37c9ba1ca37..98b4635a001 100644
--- a/video/out/x11_common.c
+++ b/video/out/x11_common.c
@@ -1608,17 +1608,16 @@ static void vo_x11_map_window(struct vo *vo, struct mp_rect rc)
         x11_send_ewmh_msg(x11, "_NET_WM_FULLSCREEN_MONITORS", params);
     }
 
-    if (x11->opts->all_workspaces || x11->opts->geometry.ws > 0) {
-        if (x11->wm_type & vo_wm_STICKY) {
-            Atom state = XA(x11, _NET_WM_STATE_STICKY);
-            XChangeProperty(x11->display, x11->window, XA(x11, _NET_WM_STATE), XA_ATOM,
-                            32, PropModeAppend, (unsigned char *)&state, 1);
-        } else {
-            long v = x11->opts->all_workspaces
-                    ? 0xFFFFFFFF : x11->opts->geometry.ws - 1;
-            XChangeProperty(x11->display, x11->window, XA(x11, _NET_WM_DESKTOP),
-                            XA_CARDINAL, 32, PropModeReplace, (unsigned char *)&v, 1);
-        }
+    if (x11->opts->geometry.ws > 0) {
+        long v = x11->opts->geometry.ws - 1;
+        XChangeProperty(x11->display, x11->window, XA(x11, _NET_WM_DESKTOP),
+                        XA_CARDINAL, 32, PropModeReplace, (unsigned char *)&v, 1);
+    }
+
+    if (x11->opts->all_workspaces && (x11->wm_type & vo_wm_STICKY)) {
+        Atom state = XA(x11, _NET_WM_STATE_STICKY);
+        XChangeProperty(x11->display, x11->window, XA(x11, _NET_WM_STATE), XA_ATOM,
+                        32, PropModeReplace, (unsigned char *)&state, 1);
     }
 
     vo_x11_update_composition_hint(vo);

```

File: video/out/x11_common.c (hunk 2 above)
Comment (odnar-dev) [reply video/out/x11_common.c:1615]: agree

File: video/out/x11_common.c (hunk 1 above)
Comment (odnar-dev) [reply video/out/x11_common.c:1618]: i sorry, but why?

Comment (sfan5) [reply video/out/x11_common.c:1618]: the current behaviour is that setting a workspace (`ws`) takes precedence over the all-workspaces option
this should be preserved

Comment (odnar-dev) [reply video/out/x11_common.c:1618]: i think the current behavior give precedence to all-workspaces option 
```c
    if (x11->opts->all_workspaces || x11->opts->geometry.ws > 0) {
        long v = x11->opts->all_workspaces
                ? 0xFFFFFFFF : x11->opts->geometry.ws - 1;
```

Comment (odnar-dev) [reply video/out/x11_common.c:1618]: i think the current behavior give precedence to all-workspaces option 
```c
    if (x11->opts->all_workspaces || x11->opts->geometry.ws > 0) {
        long v = x11->opts->all_workspaces
                ? 0xFFFFFFFF : x11->opts->geometry.ws - 1;
```

Comment (sfan5) [reply video/out/x11_common.c:1618]: oh yeah you're correct, then order needs to be swapped

**Review by sfan5 — APPROVED** [review]
Works on xfce
please squash all commits

**Comment by sfan5** [comment]
Please also remove the changelog-like entries in the commit message, they only make sense if you have seen the original commit but that now no longer exists.

**Pull request merged**